spaceapi/src/lib.rs

```diff
@@ -8,11 +8,35 @@ use rustc_serialize::json::{Json, ToJson};
 /// An Optional value can contain Optional::Some<T> or Optional::Absent.
 /// It is similar to an Option, but Optional::Absent means it will be
 /// omitted when serialized.
+#[derive(Debug, Copy, Clone)]
 pub enum Optional<T> {
     Value(T),
     Absent,
 }
 
+impl<T> Optional<T> {
+    pub fn map_or<U, F: FnOnce(T) -> U>(self, def: U, f: F) -> U {
+        match self {
+            Optional::Value(v) => f(v),
+            Optional::Absent => def,
+        }
+    }
```

Comment:
We could serialise to `null` (which btw. is the correct behaviour for objects having a key but no value). In a second step we could filter the map and replace `null` with `undefined` which will be serialised into not being present at all (at least if `ToJson` is standard compliant).

```rust
let option_tuples = options.map(|v| { (v.to_string(), v.to_json()); }));
let filtered_option_tuples = option_tuples.reject_if(|v| { v.2 == None; })
```

or similar...